### ISSUE TITLE
"pub build" remove everything in directory where it builds to 

### ISSUE THREAD

**Issue by DartBot** [dart-lang/pub#1271]
_Originally opened as dart-lang/sdk#22945_

*This issue was originally filed by hangglide.on...&#064;gmail.com*
_____
Hello, I'm new to Dart and recently I found that pub build command removes everything under directory where it writes to. For example, I have the following structure:

src/
&nbsp;&nbsp;packages/
&nbsp;&nbsp;main.dart #here I run 'pub build ../build'
build/
&nbsp;&nbsp;app.run
&nbsp;&nbsp;templates/
&nbsp;&nbsp;web/
&nbsp;&nbsp;&nbsp;&nbsp;styles/
&nbsp;&nbsp;&nbsp;&nbsp;fonts/
&nbsp;&nbsp;&nbsp;&nbsp;main.dart.js

After I get the following:

src/
&nbsp;&nbsp;main.dart
build/
&nbsp;&nbsp;web/
&nbsp;&nbsp;&nbsp;&nbsp;packages/
&nbsp;&nbsp;&nbsp;&nbsp;main.dart.js
&nbsp;&nbsp;&nbsp;&nbsp;
There is even no warning that everything under chosen directory will be destroyed. May be it isn't bug, but at least dangerous behaviour.

Related question: can I explicitly specify the folder where main.dart.js and packages/ will be written to? (I want to get exclude intermediate 'web' directory)

Ubuntu 14.10
Dart VM version: 1.9.0-dev.10.12

**Comment by DartBot** [comment]
<a href="https://github.com/zoechi"><img src="https://avatars.githubusercontent.com/u/405837?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [zoechi](https://github.com/zoechi)**

----

I can see that this is surprising for new users but after the first time you know how it works and the next x-thousend times it would become very annoying when you have to confirm the obvious every time. 

I suggest reading https://www.dartlang.org/tools/pub/package-layout.html

**Comment by DartBot** [comment]
*This comment was originally written by hangglide.o...&#064;gmail.com*
_____
Thanks, I have already read it. The most frustrating thing is that Dart's pub layout is intended to be both server and client side at the same time and there is no other option (like ecosystem, which tries to cover all that is possible). But what if what use Dart to create only client logic? Yes there is no problem to generate/build dart project to js using dart2js. But lets consider the following structure:

project/
&nbsp;&nbsp;# source files
&nbsp;&nbsp;src/
&nbsp;&nbsp;&nbsp;&nbsp;controllers/
&nbsp;&nbsp;&nbsp;&nbsp;models/
&nbsp;&nbsp;&nbsp;&nbsp;views/ # html
&nbsp;&nbsp;&nbsp;&nbsp;web/
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;styles/
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;images/
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;scripts/
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;packages/
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;main.dart
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;pubspec.yaml
&nbsp;&nbsp;&nbsp;&nbsp;main.go # I'm Go programmer

&nbsp;&nbsp;&nbsp;# here is what I want to get after build process
&nbsp;&nbsp;&nbsp;build/
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;views/
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;web/
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;styles/
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;images/
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;scripts/
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;packages/ # here we have actual files, not links
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;main.dart # copy or minified version
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;main.dart.js # the same but converted to js
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;main # compiled executable Go program

I think, the most people which decided move from JS to Dart will be happy if the build process will be simple like example above:
&nbsp;
cd src/web/scripts
dart build --minify ../../build/web/scripts # or pub build path
[process]

[there is what we get in the specified folder]
packages/ # static files
main.dart
main.dart.js
&nbsp;
I'm sorry if example too "strange/weird".

If to be honest I consider Golang the most thoughtful and minimalist language ever exists and from Go's point of view, I'm still confused why pub command named "pub". If all Dart's tools start from word dart\* and that is good, why introduce new 'pub' concept and new overhead to newbies if you could place pub functionality into dart command itself. Or if you want to keep tool of managing VM separately why you don't call it 'dartpm' or 'dartpub' and use the same name for yaml file instead weird 'pubspec.yaml'. What do you think?

**Comment by DartBot** [comment]
*This comment was originally written by hangglide...&#064;gmail.com*
_____
The example's project structure almost identical to https://github.com/google/web-starter-kit

**Comment by DartBot** [comment]
<a href="https://github.com/zoechi"><img src="https://avatars.githubusercontent.com/u/405837?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [zoechi](https://github.com/zoechi)**

----

I have troubles to understand what the problem is. You don't want to have the `web` folder below the `project` folder? 

I can tell you that you won't have much fun if you try to avoid or circumvent the pub package convention. 
All you have to do to build is running `pub build` or `pub build web` if you just want to build the web client source.

I usually have only the entry page in web and all other code below `lib` or `lib/client`, `lib/server`, and `lib/shared` if I have client and server code in one package.

I like Go and its minimalist philosophy a lot as well, but complaining why the tool is named `pub` seems your priorities should be reconsidered ;-). 
Pub is not part of the Dart language standard and there might be other implementations which provide other tools for dependency management and other functionality pub provides.

**Comment by DartBot** [comment]
*This comment was originally written by hangglide.only...&#064;gmail.com*
_____
> I have troubles to understand what the problem is.

I just want say: hey, dart take these staff (copy of packages folder, copy of main.dart and compiled main.dart.js) and place it to specified folder. However it is impossible because pub build require place dart files under one of the predefined directories and only after the process can be completed.

> Pub is not part of the Dart language standard and there might be other implementations

Now It makes sense!) I think that's all I want to hear) big thanks for answers :)

If my idea remains unclear - just close the issue.

**Comment by DartBot** [comment]
<a href="https://github.com/dgrove"><img src="https://avatars.githubusercontent.com/u/2108507?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [dgrove](https://github.com/dgrove)**

----

*Removed [Type-Defect](../labels/Type-Defect) label.*
*Added [Type-Enhancement](../labels/Type-Enhancement), [Area-Pub](../labels/Area-Pub), [Triaged](../labels/Triaged) labels.*

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

You can choose which directories "pub build" builds and where it emits its output using "pub build &lt;input> &lt;input> --output=&lt;output>". But you still need to have specific web app roots, because pub and backback need to be scoped to something, and scoping to your root directory doesn't make sense.
______
*Added [AsDesigned](../labels/AsDesigned) label.*
